feat!: pass datasource_type and datasource_id to form_data

eschutho opened the pull request:
### SUMMARY
With SIP68 we will be deprecating the ConnectorRegistry and instead having a fixed set of datasources. We currently pass both a dataset_id and datasource_type in the form_data, but we are only passing in the dataset_id to the api itself. In cases where there is no form data, we usually default to a "table" datasource, but this PR allows us to be more flexible about having different types of datasources in the future and changes the api to pass in both the datasource_type and id. We are introducing the intent to start using more datasources in [SIP81.](https://github.com/apache/superset/issues/19953)

The goal with the cache keys and temporary explore state is to be able to read from the existing format (dataset_id) but write to the new format (datasource_id and datasource_type). Any existing keys in the old format would default to a type of 'table'.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
No visual changes

### TESTING INSTRUCTIONS
Existing tests have been updated and a few new tests added to account for existing form_data cache structure. 
To test, you should be able to go through the entire explore flow without any issues. 

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

eschutho committed:
pass datasource_type and datasource_id to form_data

eschutho commented on superset/explore/form_data/commands/parameters.py:
```diff
@@ -23,7 +23,8 @@
     actor: User
-    dataset_id: int = 0
+    datasource_type: str = ""
+    datasource_id: int = 0
```
I'm not sure why we have datasource_id as required with a default to 0, so I kept with the same pattern and set datasource_type to an empty string. It almost seems like we should have two different classes here when a key exists and when it doesn't.

michael-s-molina replied:
I think it's a good idea to break it into multiple classes 👍🏼

eschutho replied:
I agree, but if it's ok, I'll put this into a different PR since it's currently one class, and outside this scope.

eschutho commented on superset/explore/permalink/commands/create.py:
```diff
@@ -39,11 +39,16 @@ def __init__(self, actor: User, state: Dict[str, Any]):
-            check_access(dataset_id, self.chart_id, self.actor)
+            datasource = self.datasource.split("__")
+            datasource_id: int = int(datasource[0])
+            datasource_type: str = datasource[1]
```
I really wanted to do a nice tuple unpacking here, but unfortunately mypy wouldn't have it.

hughhhh commented on superset/explore/form_data/commands/delete.py:
```diff
@@ -46,14 +48,15 @@ def run(self) -> bool:
                 tab_id = self._cmd_params.tab_id
                 contextual_key = cache_key(
-                    session.get("_id"), tab_id, dataset_id, chart_id
+                    session.get("_id"), tab_id, datasource_id, chart_id
```
wouldn't we need to add `datasource_type` or we would get the same cache hit for datasource_id but different datasource_type?

```suggestion
                    session.get("_id"), tab_id, datasource_id, datasource_id, chart_id
```

eschutho replied:
yes, thanks for catching!

hughhhh commented on tests/integration_tests/explore/form_data/api_tests.py:
```diff
@@ -67,65 +67,92 @@ def dataset_id() -> int:
         "form_data": INITIAL_FORM_DATA,
     }
     resp = client.post("api/v1/explore/form_data", json=payload)
+    print(resp.data)
```
remove this

eschutho replied:
ty

eschutho committed:
add datasource_type to delete command

eschutho committed:
add datasource_type to delete command

michael-s-molina commented on superset/explore/utils.py:
```diff
@@ -24,15 +24,24 @@
 
+def check_datasource_access(datasource_id: int, datasource_type: str) -> Optional[bool]:
+    if datasource_id:
+        if datasource_type == DatasourceType.TABLE:
```
I'm not sure if `check_dataset_access` depends on the `datasource_type`. Shouldn’t the check occur independently of the data source type?

eschutho replied:
Right now, it's not dependent because there's only one type, but when we start adding other types, we'll do something like this: 
```
def check_query_access(query_id: int) -> Optional[bool]:
    if query_id:
        query = QueryDAO.find_by_id(query_id)
        if query:
            can_access_datasource = security_manager.raise_for_access(
                query=query)
            if can_access_datasource:
                return True
    raise QueryNotFoundError
```

eschutho replied:
I'll leave a comment to make that more clear.

michael-s-molina commented on superset/utils/cache_manager.py:
```diff
@@ -15,15 +15,40 @@
 
+class ExploreFormDataCache(Cache):
+    def get(self, *args: Any, **kwargs: Any) -> Optional[Union[str, Markup]]:
+        cache = self.cache.get(*args, **kwargs)
```
There's a problem here with old keys because the `contextual_key` is now being generated with the datasource_type but the old keys didn't have it so it will miss the cache and return `None` for old keys.

eschutho replied:
Ok, I added a check using the old keys for update and delete as well. I don't really think there's a good way to remap these, unfortunately. The cache persistence is configurable but the default is 7 days.. maybe we can delete this fallback code after three months or so.

michael-s-molina commented on superset/explore/form_data/commands/update.py:
```diff
@@ -64,7 +65,7 @@ def run(self) -> Optional[str]:
                 tab_id = self._cmd_params.tab_id
                 contextual_key = cache_key(
-                    session.get("_id"), tab_id, dataset_id, chart_id
+                    session.get("_id"), tab_id, datasource_id, chart_id, datasource_type
```
It will miss the cache here for old keys because they were generated without the `datasource_type`. We can add logic to also query for the old format or assume that new keys will be created and think about how to clean the old entries.

eschutho replied:
This is creating a new key, though, right?

eschutho replied:
a cache key is just a uuid, right? It's not encrypted data?

michael-s-molina commented on tests/integration_tests/explore/form_data/api_tests.py:
```diff
@@ -67,65 +67,91 @@ def dataset_id() -> int:
         return dataset.id
 
 
+@pytest.fixture
```
Maybe have one fixture that returns the `datasource` with both `id` and `type`?

eschutho replied:
yeah, good call

michael-s-molina commented on tests/integration_tests/utils/cache_manager_tests.py:
```diff
@@ -0,0 +1,50 @@
+
+from superset.extensions import cache_manager
+from superset.utils.core import DatasourceType
+
```
Can we add a test that checks for retrieval using keys that were generated without a datasource type?

eschutho replied:
yeah, got it on line 38. 👍

michael-s-molina reviewed (commented):
Thanks for the PR @eschutho! I left some comments as a result of a first-pass review.

eschutho commented on superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx:
```diff
@@ -189,9 +189,10 @@ class DatasourceControl extends React.PureComponent {
+      const datasourceType = getUrlParam(URL_PARAMS.datasourceType);
       const sliceId = getUrlParam(URL_PARAMS.sliceId);
-      if (!datasetId && !sliceId) {
+      if (!datasourceId && !sliceId && !datasourceType) {
```
Thanks, sounds good. I think we set a default for type then we won't be able to validate if it's missing.

eschutho commented on superset-frontend/src/constants.ts:
```diff
@@ -84,7 +88,8 @@ export const URL_PARAMS = {
   URL_PARAMS.sliceId.name,
-  URL_PARAMS.datasetId.name,
+  URL_PARAMS.datasourceId.name,
+  URL_PARAMS.datasourceType.name,
```
@michael-s-molina this could be an old URL, too? I'll datasetId back in here..

eschutho commented:
Thanks for the review and added context @michael-s-molina. I made some small changes based on the feedback, but still have to come back to this and add in the check access query logic that I removed.

john-bodley commented:
@eschutho I'm a little lost, why do we need the `datasource_type` post the Druid NoSQL deprecation? i.e., this served merely to ensure the (`datasource_type`, `datasouce_id`) tuple was unique but given we only now have `datasource_type = 'table` this is irrelevant isn't it—unless this is somehow being used @betodealmeida's table/query schema refactor.

eschutho commented:
> @eschutho I'm a little lost, why do we need the `datasource_type` post the Druid NoSQL deprecation? i.e., this served merely to ensure the (`datasource_type`, `datasouce_id`) tuple was unique but given we only now have `datasource_type = 'table` this is irrelevant isn't it—unless this is somehow being used @betodealmeida's table/query schema refactor.

yeah, this is a precursor for https://github.com/apache/superset/issues/19953 where we'll be introducing other datasource types: tables, queries, savedqueries.

hughhhh commented on superset/utils/cache_manager.py:
```diff
@@ -15,15 +15,40 @@
+            return None
+
+        # rename data keys for existing cache based on new TemporaryExploreState model
+        if isinstance(cache, dict):
```
Why do we need to check if the cache is a dict? Won't it always be a dict?

eschutho replied:
actually, some times it returns a string.

eschutho replied:
superset/explore/form_data/commands/create.py line 52 for example, it returns the key, not the data. I'm honestly not sure why.

jinghua-qa commented:
/testenv up

eschutho commented on scripts/tests/run.sh:
```diff
@@ -138,5 +138,5 @@ fi
 if [ $RUN_TESTS -eq 1 ]
 then
-  pytest --durations=0 --maxfail=1 "${TEST_MODULE}"
+  pytest --durations=0 --maxfail=10 "${TEST_MODULE}"
```
```suggestion
  pytest --durations=0 --maxfail=1 "${TEST_MODULE}"
```

eschutho committed:
fix old keys implementation

eschutho commented on superset/explore/form_data/commands/delete.py:
```diff
@@ -47,14 +49,15 @@ def run(self) -> bool:
                 tab_id = self._cmd_params.tab_id
                 contextual_key = cache_key(
-                    session.get("_id"), tab_id, dataset_id, chart_id
+                    session.get("_id"), tab_id, datasource_id, chart_id, datasource_type
```
todo: I still need to write up something here for deleting old keys. Unless it's ok that it just expires on it's own. The user will get a 404 though.

hughhhh replied:
Since we are generating new keys for misses, i think it's fine to let the keys expire. 

I think it would be a lot of work to have another layer of remapping old keys to new keys.

eschutho replied:
Yeah, let's do that, it's only for delete and update. Unless anyone else has thoughts.

eschutho commented on tests/integration_tests/explore/form_data/commands_tests.py:
```diff
@@ -0,0 +1,130 @@
+        assert isinstance(command.run(), str)
+
+
+# TODO
```
I'd like to finish writing up these tests, too.

eschutho commented:
This is mostly done.. I just found one other thing last minute that I'd like to fix. I'll still welcome any feedback.

hughhhh reviewed (approved):
🚢

2 things i want to call out though is having a regression testing to make sure this works intended and we aren't deploying an huge bugs into the explore view.

Also is possible for us to add some statsd metrics to track hit vs. misses with the cache. I think this will help us understand how the logic is effecting the end user with pulling data. (this can be another ticket)

eschutho committed:
add more tests

eschutho commented:
Thanks @jinghua-qa, I didn't know about that feature! I checked it on my local and then on the ephemeral and I see the highlighting.. is this the expected behavior?

https://user-images.githubusercontent.com/5186919/171676575-a5d18f6e-cdc3-4a01-be32-4b5889184ca9.mov

hughhhh commented:
🚢 🚢 🚢
